### PR TITLE
close #951 include keyName prop for custom id

### LINKED ISSUE #951: custom ID with useFieldArray doesn't work
**Describe the bug**
When using useFieldArray, if an array field has custom ID set then dirty state detection doesn't work.

**To Reproduce**
1. Go to https://codesandbox.io/s/react-hook-form-dirty-state-with-custom-id-kpq7f
2. Click on append and then remove the appended item
3. Observe that formState.dirty is still false.

**Additional context**
 - Version 4.8.0

This works fine if the code that sets the initial field IDs is removed.

### PR DESCRIPTION
**Pull request by bluebill1049**
issue: https://github.com/react-hook-form/react-hook-form/issues/951

```
useFieldArray({
  control,
  kayName: 'id', // default as id and you can change to others
  name: 'test'
})
```

### CODE REVIEW

**Review by kotarella1110 — COMMENTED** [review]
I tried to improve the type to be generated automatically.
During trying, it was found that if any one of several generic parameters was given a type, the rest generic parameters would not be inferred as follows.
Please see this [Issue](https://github.com/microsoft/TypeScript/issues/26242) and [PR](https://github.com/microsoft/TypeScript/pull/26349) for the details.

playground: http://www.typescriptlang.org/play/?ssl=12&ssc=42&pln=12&pc=51#code/GYVwdgxgLglg9mABFApgZygHgLAChEGICCiKAHqmACZqIYBOMYA5ogLx1SMsA0ehiAEKkKKarTAgAtgCMU9dokmz5ffIQDCIyjUQy4cADYoAhkg76jpsHgB8AChMAuYjz0vBbjS40BKD4gA3vyE9ChQIPRIMgDceAC+eHioGPYA5CZpbgCMblwgKL4xiAD0JYhMwPJhVIj2EHD0YdC+eGWIAKJkAA4o0Ci1UACevS6gkLAIyOhYAEQmszl59AUOzojzi+6IuYjeyCuFLtlt5QBKfSgwAG4DyCMoY+DQ8EgpcwtLB6uOLptuMmOXhc+SOO2SM0wGTSDmhX1BRVK5TAcAqYCqTTu9iYDUxLVOnR6fVQgweTwmr2mGEw-x2yx+61pgLpexBh384Nw7QuECut1Jo0Q4xeU3eNM+SmkcnoAIMxjMiFsdUZEuZymlwL0cusHPV8iAA

```ts
function test<
    A extends string = string,
    B extends number = number,
    C extends boolean = boolean
>(a: A, b: B, C: C): B {
    return b;
}

test('a', 1, true); // inferred (correct)
// Expected type: function test<"a", 1, true>(a: "a", b: 1, C: true): 1
// Received type: function test<"a", 1, true>(a: "a", b: 1, C: true): 1
test<'a'>('a', 1, true); // no inferred (incorrect)
// Expected type: function test<"a", 1, true>(a: "a", b: 1, C: true): 1
// Received type: function test<"a", number, boolean > (a: "a", b: number, C: boolean): number
```

Therefore, the type could not be generated automatically.

codesandbox: https://codesandbox.io/s/react-hook-form-usefieldarray-pr-957-9tmp3

```ts
useFieldArray<{ name: string }>({
  keyName: "customId", // type error
  control,
  name: "test"
});
```

However, we can get the correct type by currying.

codesandbox: https://codesandbox.io/s/react-hook-form-usefieldarray-pr-957-curried-xpnfx

```ts
useFieldArray<{ name: string }>()({
  keyName: "customId",
  control,
  name: "test"
});
```

File: app/src/useFieldArray.tsx [hunk 1]
```diff
@@ -35,7 +35,7 @@ const UseFieldArray: React.FC = (props: any) => {
     move,
     insert,
     remove,
-  } = useFieldArray<{ name: string }>({
+  } = useFieldArray<{ name: string, id: string }>({
```
[the remark this answers is not in the document]
Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: i agree, do you have any suggestion @kotarella1110

[the remark this answers is not in the document]
Comment (kotarella1110) [reply app/src/useFieldArray.tsx]: @bluebill1049 Suggest the following:

1. Only when setting `keyName`, specify the generics parameter of `keyName`

```ts
const { fields } = useFieldArray<{ name: string }, "customId">()({
  keyName: "customId",
  control,
  name: "data"
});
// const fields: Partial<{
//     name: string;
//     customId: string
// }>[]

useFieldArray<{ name: string }>()({
  control,
  name: "data"
});
// const fields: Partial<{
//     name: string;
//     id: string
// }>[]
```

2. currying

```ts
useFieldArray<{ name: string }>()({
  keyName: "customId",
  control,
  name: "test"
});
```

[the remark this answers is not in the document]
Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: let's go with option 1. you want to commit on this branch?

[the remark this answers is not in the document]
Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: by the way we almost reaching 1Million downloads 🥂

[the remark this answers is not in the document]
Comment (kotarella1110) [reply app/src/useFieldArray.tsx]: Let's merge this PR once.
I will create the new PR.

[the remark this answers is not in the document]
Comment (bluebill1049) [reply app/src/useFieldArray.tsx]: yea better that way, we need more commits from you! hahah

[the remark this answers is not in the document]
Comment (kotarella1110) [reply app/src/useFieldArray.tsx]: I will do it today!

**Pull request merged**